handler.go

```diff
@@ -0,0 +1,318 @@
+	if err != nil {
+		r.out.Fatalf("Failed to create streaming request: %v\n", err)
+	}
+	streamReq.Request, err = prepareRequest(streamReq.Request, r.headers, r.opts)
+	if err != nil {
+		r.out.Fatalf("Failed while preparing the request: %v\n", err)
+	}
+
+	requestSupplier := func() ([]byte, error) {
+		msg, err := streamMsgReader.NextBody()
+		if err == io.EOF {
+			return nil, err
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Failed while reading stream input: %v", err)
+		}
+		return msg, nil
+	}
+
+	responseHandler := func(body []byte) error {
+		res, err := r.serializer.Response(&transport.Response{Body: body})
+		if err != nil {
+			return fmt.Errorf("Failed while serializing stream response: %v", err)
+		}
+
+		bs, err := json.MarshalIndent(res, "", "  ")
+		if err != nil {
+			return fmt.Errorf("Failed to convert map to JSON: %v\nMap: %+v", err, res)
+		}
+		r.out.Printf("%s\n\n", bs)
+		return nil
+	}
+
+	if r.shouldMakeInitialRequest() {
+		err = makeStreamRequest(r.transport, streamReq, streamSerializer, requestSupplier, responseHandler)
+		if err != nil {
+			r.out.Fatalf("%v\n", err)
+		}
+	}
+}
+
+// shouldMakeInitialRequest returns true if initial request must be made
+func (r requestHandler) shouldMakeInitialRequest() bool {
+	return !(r.opts.BOpts.enabled() && r.opts.BOpts.WarmupRequests == 0)
+}
+
+// isStreamingMethod returns true if RPC is streaming type
+func isStreamingMethod(serializer encoding.Serializer) bool {
+	streamSerializer, ok := serializer.(encoding.StreamSerializer)
+	if !ok {
+		return false
+	}
+	return streamSerializer.MethodType() != encoding.Unary
+}
+
+// makeStreamRequest creates a stream from the given transport and handles
+// the stream request and response
+func makeStreamRequest(t transport.Transport, streamReq *transport.StreamRequest, serializer encoding.StreamSerializer,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+	streamTransport, ok := t.(transport.StreamTransport)
+	if !ok {
+		return fmt.Errorf("Transport does not support stream calls: %q", t.Protocol())
+	}
+
+	ctx, cancel := tchannel.NewContext(streamReq.Request.Timeout)
+	defer cancel()
+	ctx = makeContextWithTrace(ctx, t, streamReq.Request, 0)
+
+	stream, err := streamTransport.CallStream(ctx, streamReq)
+	if err != nil {
+		return fmt.Errorf("Failed while making stream call: %v", err)
+	}
+
+	rpcType := serializer.MethodType()
+	if rpcType == encoding.BidirectionalStream {
+		return makeBidiStream(ctx, stream, reqSupplier, resHandler)
+	} else if rpcType == encoding.ClientStream {
+		return makeClientStream(ctx, stream, reqSupplier, resHandler)
+	}
+	return makeServerStream(ctx, stream, reqSupplier, resHandler)
+}
+
+// makeServerStream starts server-side streaming rpc
+func makeServerStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	reqSupplier streamRequestSupplier, resHandler streamResponseHandler) error {
+
+	req, err := reqSupplier()
+	if err != nil && err != io.EOF {
```

Comment:
Yes, this is a server-side streaming method where client sends the initial mandatory request and server streams the response. In case user does not provide any input we send nil request, this is similar to the unary request body handling in Yab

Reply:
Add a comment please